packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,134 @@ Thank you for taking the time to contribute.
+
+Lower-level components, which are usually composed together to build the second type of components: UI Blocks. UI Primitives usually don't contain translated text and aim at being highly configurable (at the expense of exposing more props and therefore more complexity).
+
+UI Primitives can be useful when in need of building an app with all new UI that looks and feels like WordPress but isn't powered by other WordPress technologies (like `@wordpress/i18n`).
+
+_[Note: we need to discuss how to properly classify/label these components. Should we introduce `primitives`, `blocks` and `utils` folders in the `src` folder? Should we include a standard way in each component's README to show which category it belongs to?]_
+
+#### UI Blocks
+
+Higher-level components, which usually offer fewer props compared to UI Primitives and contain translated text. These components use other WordPress technologies and are ready to be used in Gutenberg's UI.
+
+#### Utilities
+
+Components and other utilities function that don't necessarily render a piece of UI to screen, but are instead useful when implementing a given piece of functionality.
+
+### Components composition
+
+[To be expanded] E.g.:
+
+- Polymorphic Components
+- Using `children` vs custom render props vs arbitrary "data" props
+- Controlled and semi-controlled components
+- Composition patterns
+- Sub-components naming conventions
+- Components' layout responsibilities and boundaries (i.e., a component should only affect the layout of its children, not its own)
+- ...
+
+### APIs Consinstency
+
+[To be expanded] E.g.:
+
+- Boolean component props should be prefixed with `is*` (e.g. `isChecked`), `has*` (e.g. `hasValue`) or `enable*` (e.g. `enableScroll`)
+- Event callback props should be prefixed with `on*` (e.g. `onChanged`)
+- ...
+
+### Technical requirements for new components
+
+The following are a set of technical requirements for all newly introduced components. These requirements are also retroactively being applied to existing components.
+
+For an example of a component that follows these requirements, take a look at [`ItemGroup`](/packages/components/src/item-group);
+
+#### TypeScript
+
+All new components should be written in TypeScript and should be typed using the `PolymorphicComponent` type (more details about polymorphism can be found above in the "Components composition" section).
+
+#### Emotion
+
+All new component should be styled using [Emotion](https://emotion.sh/docs/introduction).
+
+Note: Instead of using Emotion's standard `cx` function, the custom [`useCx` hook](/packages/components/src/utils/hooks/use-cx.ts) should be used instead.
+
+#### Context system
+
+[To be expanded]
+
+#### Hooks vs Components
+
+[To be expanded]
+
+#### Unit tests
+
+All new components should include unit tests using [testing library](https://testing-library.com/). Ideally, the component's main piece of functionality would be tested explicitely (rather than using snapshots).
```

Comment:
Is it really necessary to have mandatory unit tests for components. Usually they turned out to be testing implementation details.

Reply:
> Usually they turned out to be testing implementation details.

I agree that most unit tests are either snapshots or usually poorly written. Ideally, it should be the reviewer's duty to ensure that the unit tests added to the codebase are well written.

Is not requesting unit tests a better solution?

Reply:
Update: I merged #34423, rebased this PR, and updated the "Unit Tests" section in 0a36f58 (#33960) as agreed in [this conversation](https://github.com/WordPress/gutenberg/pull/33960#discussion_r690196895).

There's no mention of "required" unit tests anymore, but I'd be interested to know your opinion on my answer above.